plugin/better-whitespace.vim

```diff
@@ -71,113 +85,106 @@ endif
-            let g:current_line_whitespace_disabled_soft = 1
-            let g:current_line_whitespace_disabled_hard = 0
-            call s:InAllWindows("match ExtraWhitespace ''")
-            call <SID>Echo("Current Line Highlight Off (soft)")
+" Setup matching with either syntax or match
+if g:current_line_whitespace_disabled_soft == 1
+    " Match Whitespace on all lines
+    function! s:HighlightEOLWhitespace()
+        call <SID>ClearHighlighting()
+        if <SID>ShouldHighlight()
+            exe 'syn match ExtraWhitespace excludenl "' . s:eol_whitespace_pattern . '"'
         endif
-        " Re-run auto commands with the new settings
-        call <SID>SetupAutoCommands()
-    endif
-endfunction
+    endfunction
 
-" Enables whitespace highlighting for the current line
-function! s:CurrentLineWhitespaceOn()
-    if g:better_whitespace_enabled == 1
-        let g:current_line_whitespace_disabled_hard = 0
-        let g:current_line_whitespace_disabled_soft = 0
-        call <SID>SetupAutoCommands()
-        call s:InAllWindows('syn clear ExtraWhitespace | match ExtraWhitespace "' . s:eol_whitespace_pattern . '"')
-        call <SID>Echo("Current Line Highlight On")
-    endif
+    " Match Whitespace on all lines except the current one
+    function! s:HighlightEOLWhitespaceExceptCurrentLine()
+        call <SID>ClearHighlighting()
+        if <SID>ShouldHighlight()
+            exe 'syn match ExtraWhitespace excludenl "\%<' . line('.') .  'l' . s:eol_whitespace_pattern .
+                                                 \ '\|\%>' . line('.') .  'l' . s:eol_whitespace_pattern . '"'
+        endif
+    endfunction
+
+    " Remove Whitespace matching
+    function! s:ClearHighlighting()
+        syn clear ExtraWhitespace
+    endfunction
+else
+    " Match Whitespace on all lines
+    function! s:HighlightEOLWhitespace()
+        if <SID>ShouldHighlight()
+            exe 'match ExtraWhitespace "' . s:eol_whitespace_pattern . '"'
+        else
+            call <SID>ClearHighlighting()
+        endif
+    endfunction
+
+    " Match Whitespace on all lines except the current one
+    function! s:HighlightEOLWhitespaceExceptCurrentLine()
+        if <SID>ShouldHighlight()
+            exe 'match ExtraWhitespace "\%<' . line('.') .  'l' . s:eol_whitespace_pattern .
+                                   \ '\|\%>' . line('.') .  'l' . s:eol_whitespace_pattern . '"'
+        else
+            call <SID>ClearHighlighting()
+        endif
+    endfunction
+
+    " Remove Whitespace matching
+    function! s:ClearHighlighting()
+        match ExtraWhitespace ''
+    endfunction
+endif
+
+" Checks for extraneous whitespace in the file
+" WARNING: moves the cursor.
+function! s:DetectWhitespace(line1, line2)
+    call cursor(a:line1, 1)
+    return search(s:strip_whitespace_pattern, 'cn', a:line2)
 endfunction
 
 " Removes all extraneous whitespace in the file
-function! s:StripWhitespace( line1, line2 )
+function! s:StripWhitespace(line1, line2)
```

Comment:
It might be nice to give an error (`E21`) if `modifiable` isn't set when this is invoked. This would be consistent with what you get when trying to execute a Vim built-in on a non-modifiable buffer. As-is, this will just appear as though it ran normally to the user.

Reply:
The strip whitespace on save defaults to `off` when `modifiable` is set, which I think is sensible. I agree that an error instead of a nop would be more useful if this function is then called, e.g. manually, but this is hidden by the `silence!` that we use.

Error handling on the whole could be a separate PR maybe? The best way to do this imho would be to check `v:errmsg` and print it if it's not `E486: Pattern not found`.

Reply:
Went with the latter option.